---
title: 第十三节：异常
tags:
  - 课后解答
url: 245.html
id: 245
categories:
  - 学习
  - 教学

date: 2018-05-28 14:36:29
---

\[success\]  

一：异常及其异常处理：
-----------

  **什么是异常？**   ------异常：不正常的情况

*   异常即是一个事件，该事件会在程序执行过程中发生，影响了程序的正常执行。一般情况下，在Python无法正常处理程序时就会发生一个异常。
*   异常是Python对象，表示一个错误。
*   当Python脚本发生异常时我们需要捕获处理它，否则程序会终止执行。

* * *

**异常出现的原因：** 不正常的情况，在程序中，会有两种体现：

*   代码错误或语法错误；程序都运行不起来！
*   程序运行过程中，在某些特定条件下，不合适的数据引起程序出现错误导致程序崩溃

> 例如：_要求用户输入一个数字，但是用户误操作输入了字符串，在进行类型转换时就会出现错误导致程序崩溃！_

* * *

**为什么要处理异常？** 当程序在运行过程中，由于用户的误操作或者不合适的数据引发的程序错误，让代码自己处理并保证程序的正常执行。而不至于因为错误导致程序崩溃！        这样提高代码的健壮性！

* * *

**怎么处理异常？** l 捕获处理异常 l 断言处理异常【测试异常信息】 **捕获处理异常语法：** 捕捉异常可以使用 try/except 语句。 try/except语句用来检测try语句块中的错误，从而让except语句捕获异常信息并处理。 如果你不想在异常发生时结束你的程序，只需在try里捕获它。 语法： 以下为简单的_try....except...else_的语法：

![复制代码](https://common.cnblogs.com/images/copycode.gif)

 1 try:
 2 
 3 <语句\>        #运行别的代码
 4 
 5 except <名字>：
 6 
 7 <语句\>        #如果在try部份引发了'name'异常
 8 
 9 except <名字>，<数据>: 10 
11 <语句\>        #如果引发了'name'异常，获得附加的数据
12 
13 else: 14 
15 <语句\>        #如果没有异常发生
16 
17 finally： 18 
19 <语句\>                #无论try语句是否发生异常，都要执行的语句



**处理异常：** **1\. 基本异常处理** try-except直接处理异常【可以处理任何异常—不能定位具体是什么异常】 try-except \[异常信息\] 【可以处理指定的异常】 **2\. 常见的异常** 所有异常的超类：BaseException 标准异常的超类：Exception(BaseException) 程序执行过程中经常看到的异常： …… **3\. 处理多个异常** try-except (E1, E2, E3, …,En) as **e** 一个except处理指定的多个异常 try-except E1-except E2-except E3… 每个except处理指定的异常 通常情况下，我们要查看异常信息，就需要打印异常信息—打印**e**变量的值 问题： 1） 系统提供的异常信息，晦涩难懂！ 2） 系统提供的异常信息，毕竟有限！ **4\. 自定义异常** 通过创建一个新的异常类，程序可以命名它们自己的异常。异常应该是典型的继承自Exception类，通过直接或间接的方式。 以下为与RuntimeError相关的实例,实例中创建了一个类，基类为RuntimeError，用于在异常触发时输出更多的信息。 在try语句块中，用户自定义的异常后执行except块语句，变量 e 是用于创建Networkerror类的实例。

1 class Networkerror(RuntimeError): 2 
3 def \_\_init\_\_(self, arg): 4 
5 self.args = arg

在你定义以上类后，你可以触发该异常，如下所示：


'''
 6 try:
 7 
 8 　　raise Networkerror("Bad hostname")
 9 
10 except Networkerror,e: 11 
12　　 print e.args
'''
#### 异常的参数

一个异常可以带上参数，可作为输出的异常信息参数。 **5\. 主动抛出异常** 我们可以使用raise语句自己触发异常 raise语法格式如下：  

raise \[Exception \[, args \[, traceback\]\]\]

语句中Exception是异常的类型（例如，NameError）参数是一个异常参数值。该参数是可选的，如果不提供，异常的参数是"None"。 最后一个参数是可选的（在实践中很少使用），如果存在，是跟踪异常对象。

##### 实例

一个异常可以是一个字符串，类或对象。 Python的内核提供的异常，大多数都是实例化的类，这是一个类的实例的参数。 定义一个异常非常简单，如下所示：

def functionName( level ):

if level < 1: 
raise "Invalid level!", level 5         
The code below to this would not be executed
if we raise the exception 

  **注意：**为了能够捕获异常，"except"语句必须有用相同的异常来抛出类对象或者字符串。 例如我们捕获以上异常，"except"语句如下所示：

7     try: 8 Business Logic here... 9     except "Invalid level!": 10       Exception handling here... 11    else: 12       Rest of the code here...

在开发的过程中，主动出现一种错误，将错误抛出给程序告诉程序出错了。 **6.异常处理理解** 自定义异常，就是为了专门抛出错误的，抛出错误，就是严重警告这里出现了什么问题 首先-代码执行过程中，出现了异常【系统标准异常】【信息不是很明白】 捕获系统异常，创建一个自定义异常 抛出自定义异常【自定义异常】【信息明确的错误】 自定义异常的目的：转换异常信息，将不明确的异常信息转换成更加精确的异常信息 **转换异常信息：异常的传递~传递更加明确的异常，给后面的代码进行处理！**  

7.Python标准异常
------------

异常名称

描述

BaseException

所有异常的基类

SystemExit

解释器请求退出

KeyboardInterrupt

用户中断执行(通常是输入^C)

Exception

常规错误的基类

StopIteration

迭代器没有更多的值

GeneratorExit

生成器(generator)发生异常来通知退出

StandardError

所有的内建标准异常的基类

ArithmeticError

所有数值计算错误的基类

FloatingPointError

浮点计算错误

OverflowError

数值运算超出最大限制

ZeroDivisionError

除(或取模)零 (所有数据类型)

AssertionError

断言语句失败

AttributeError

对象没有这个属性

EOFError

没有内建输入,到达EOF 标记

EnvironmentError

操作系统错误的基类

IOError

输入/输出操作失败

OSError

操作系统错误

WindowsError

系统调用失败

ImportError

导入模块/对象失败

LookupError

无效数据查询的基类

IndexError

序列中没有此索引(index)

KeyError

映射中没有这个键

MemoryError

内存溢出错误(对于Python 解释器不是致命的)

NameError

未声明/初始化对象 (没有属性)

UnboundLocalError

访问未初始化的本地变量

ReferenceError

弱引用(Weak reference)试图访问已经垃圾回收了的对象

RuntimeError

一般的运行时错误

NotImplementedError

尚未实现的方法

SyntaxError

Python 语法错误

IndentationError

缩进错误

TabError

Tab 和空格混用

SystemError

一般的解释器系统错误

TypeError

对类型无效的操作

ValueError

传入无效的参数

UnicodeError

Unicode 相关的错误

UnicodeDecodeError

Unicode 解码时的错误

UnicodeEncodeError

Unicode 编码时错误

UnicodeTranslateError

Unicode 转换时错误

Warning

警告的基类

DeprecationWarning

关于被弃用的特征的警告

FutureWarning

关于构造将来语义会有改变的警告

OverflowWarning

旧的关于自动提升为长整型(long)的警告

PendingDeprecationWarning

关于特性将会被废弃的警告

RuntimeWarning

可疑的运行时行为(runtime behavior)的警告

SyntaxWarning

可疑的语法的警告

UserWarning

用户代码生成的警告

二：python断言
----------

使用assert断言是学习**[python](http://www.iplaypy.com/ "python下载")**一个非常好的习惯，python **assert 断言**句语格式及用法很简单。在没完善一个程序之前，我们不知道程序在哪里会出错，与其让它在运行最崩溃，不如在出现错误条件时就崩溃，这时候就需要assert断言的帮助。本文主要是讲assert断言的基础知识。

python assert断言的作用
------------------

python assert断言是声明其布尔值必须为真的判定，如果发生异常就说明表达示为假。可以理解assert断言语句为raise-if-not，用来测试表示式，其返回值为假，就会触发[异常](http://www.iplaypy.com/jichu/exception.html)。

assert断言语句的语法格式
---------------

assert python 怎么用？ expression assert 表达式 下面做一些assert用法的语句供参考：
assert 1==1 
assert 2+2==2*2 
assert len(\['my boy',12\])<10
assert range(4)==\[0,1,2,3\]

如何为assert断言语句添加异常参数
-------------------

assert的异常参数，其实就是在断言表达式后添加[字符串](http://www.iplaypy.com/jichu/str.html)信息，用来解释断言并更好的知道是哪里出了问题。格式如下： assert expression \[, arguments\] 
assert 表达式 \[, 参数\]
